system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -833,6 +848,346 @@ extern ICompiledXpath* compileXpath(const char * xpath)
+        {
+            for (int i=0; i<eval->nodesetval->nodeNr; i++)
+            {
+                xmlNodePtr node = eval->nodesetval->nodeTab[i];
+                if (node==nullptr)
+                  continue;
+                const char *name = queryAttrValue(node, "name");
+                if (xmlHasProp(node,(const xmlChar *) "select"))
+                  tgtXpathCtx->addInputXpath(name, queryAttrValue(node,"select"));
+                else
+                  tgtXpathCtx->addInputValue(name, queryAttrValue(node,"value"));
+            }
+        }
+        xmlXPathFreeObject(eval);
+    }
+    const char *getXPathString(const char *xpath, StringBuffer &s) const override
+    {
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) xpath, xpathCtx);
+        if (eval)
+        {
+            xmlChar *v = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (v)
+            {
+                if (*v)
+                    s.append((const char *)v);
+                xmlFree(v);
+                return s;
+            }
+        }
+        return nullptr;
+    }
+    __int64 getXPathInt64(const char *xpath, __int64 dft=0) const override
+    {
+        __int64 ret = dft;
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) xpath, xpathCtx);
+        if (eval)
+        {
+            xmlChar *val = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (val)
+            {
+                if (*val)
+                    ret = _atoi64((const char *)val);
+                xmlFree(val);
+            }
+        }
+        return ret;
+    }
+    bool getXPathBool(const char *xpath, bool dft=false) const override
+    {
+        bool ret = dft;
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) xpath, xpathCtx);
+        if (eval)
+        {
+            xmlChar *val = xmlXPathCastToString(eval);
+            xmlXPathFreeObject(eval);
+            if (val)
+            {
+                if (*val)
+                    ret = strToBool((const char *)val);
+                xmlFree(val);
+            }
+        }
+        return ret;
+    }
+
+    xmlNodePtr getSection(const char *name)
+    {
+        return getSectionNode(name, nullptr);
+    }
+    xmlNodePtr ensureSection(const char *name)
+    {
+        xmlNodePtr sect = getSection(name);
+        if (sect)
+            return sect;
+        return xmlNewChild(root, nullptr, (const xmlChar *) name, nullptr);
+    }
+    void removeSection(const char *name)
+    {
+        xmlXPathObjectPtr eval = xmlXPathEval((const xmlChar *) name, xpathCtx);
+        if (!eval)
+            return;
+        if (XPATH_NODESET==eval->type && !xmlXPathNodeSetIsEmpty(eval->nodesetval))
+        {
+            xmlNodeSetPtr ns = eval->nodesetval;
+            int count = xmlXPathNodeSetGetLength(ns);
+            for(int i=0; i<count; i++)
```

Comment:
Based on a comment in the example at http://www.xmlsoft.org/examples/xpath2.c, should the loop traverse in the opposite order? The evaluation could return a node and that node's descendant, and freeing the first node frees the descendant, invalidating the node reference.

Reply:
Agreed.